### PR TITLE
Deprecate GlobalQos and delayed exchange plugin for RabbitMQ 4.3+

### PR DESCRIPTION
**Pull request by niemyjski**
## Summary

- Mark `GlobalQos` property and builder method as `[Obsolete]` since RabbitMQ 4.3+ moved `global_qos` to `denied_by_default`
- Guard `BasicQosAsync`: when connected to RabbitMQ 4.3+, downgrade to per-channel prefetch (`global: false`) with a warning instead of sending `global: true` which would kill the channel
- Upgrade delayed exchange plugin warnings to clearly state that delayed messages fall back to in-memory scheduling and that plugin support will be removed in a future version

## Context

RabbitMQ [4.3.0](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v4.3.0) (released April 2026) removed Mnesia entirely and moved several deprecated features to `denied_by_default`. The latest patch is [4.3.1](https://github.com/rabbitmq/rabbitmq-server/releases/tag/v4.3.1).

Affected features:
- `global_qos` - server rejects `BasicQos(global: true)` unless explicitly re-enabled via `deprecated_features.permit.global_qos = true` in `rabbitmq.conf`
- `rabbitmq_delayed_message_exchange` plugin - depends on Mnesia, incompatible with 4.3+

This library already detected 4.3+ and skipped the delayed exchange probe, but the `GlobalQos` option had no guard and would cause channel death on 4.3+ servers. Neither feature was marked deprecated in the public API.

## Test plan

- [x] `dotnet build` passes with 0 warnings, 0 errors
- [ ] Verify `[Obsolete]` compiler warning appears when consumer code uses `GlobalQos`
- [ ] Verify on RabbitMQ 4.3+ server: no channel errors when `GlobalQos = true` is configured (graceful downgrade with warning)
- [ ] Verify delayed message in-memory fallback still works correctly

### CODE REVIEW

**Review by niemyjski — COMMENTED** [review]
## Self-Review: PR #78 — Deprecate GlobalQos and delayed exchange plugin for RabbitMQ 4.3+

Self-review completed — issues found. Action items:
1. **WARNING (naming):** `_delayedExchangePluginIncompatibleVersion` is reused for the GlobalQos version gate (line 150). These are conceptually independent deprecations that happen to share the same threshold today. Consider introducing a separate `_globalQosRemovedVersion` constant to avoid future coupling bugs if thresholds diverge.
2. **WARNING (false deprecation warning on unknown servers):** When `_serverVersion` is null (unparseable), the `else if (useGlobalQos)` branch emits a deprecation warning ("deprecated in RabbitMQ 4.3+") even though the server may be pre-4.3. Consider gating this warning behind `_serverVersion is not null` or adjusting the message to say "may be deprecated" when version is unknown.

<details>
<summary>Full Review Details</summary>

### Security: PASS
No credential handling, no build script modifications, no obfuscated code, no new dependencies.

### Build: PASS | Tests: all CI green | CI: pass (build x2, CodeQL, CLA)
- Local build: 0 warnings, 0 errors (net8.0 + net10.0)
- Local tests fail only due to no local RabbitMQ (connection refused) — infrastructure-dependent integration tests
- CI builds all pass

### Dependencies: N/A
No dependency changes in this PR.

### Requirements: ALIGNED
- ✅ `[Obsolete]` on `GlobalQos` property and builder method
- ✅ Runtime guard: `global: true` is downgraded to `false` on 4.3+ servers
- ✅ Deprecation warnings emitted for both pre-4.3 (advisory) and 4.3+ (active downgrade)
- ✅ Delayed exchange log levels upgraded from Info/Trace to Warning/Debug
- ✅ Backward compatibility preserved — existing code compiles (with warning) and behaves correctly

### Code Quality
**Good:**
- Pragma suppression pattern is clean and well-commented
- Log messages are structured and actionable
- No breaking behavioral changes — graceful degradation
- Version detection reuses existing infrastructure

**Findings:**
1. `_delayedExchangePluginIncompatibleVersion` semantic reuse — two unrelated features sharing one version constant
2. Deprecation warning fires even when `_serverVersion` is null (unparseable) — users on servers where version detection fails get a possibly incorrect deprecation message
3. Minor: PR body says "silently downgrade" but a `LogWarning` is emitted (not silent) — cosmetic mismatch between docs and behavior

### PR-Level
- No breaking changes to public API (additive `[Obsolete]` attribute)
- No new tests added for the version-gated behavior — acceptable for now since the logic is simple branching, but a unit test for `ParseServerVersion` + version comparison edge cases would strengthen confidence
- Commit history is clean (single logical change)

</details>

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 1]
```diff
@@ -144,9 +144,22 @@ protected override async Task EnsureTopicSubscriptionAsync(CancellationToken can
             // Set QoS (Quality of Service) settings for the consumer
             if (_options.PrefetchCount > 0 || _options.PrefetchSize > 0)
             {
-                await _subscriberChannel.BasicQosAsync(_options.PrefetchSize, _options.PrefetchCount, _options.GlobalQos, cancellationToken).AnyContext();
+#pragma warning disable CS0618 // GlobalQos is obsolete but we still need to read it for backward compatibility
+                bool useGlobalQos = _options.GlobalQos;
+#pragma warning restore CS0618
+                if (useGlobalQos && _serverVersion is not null && _serverVersion >= _delayedExchangePluginIncompatibleVersion)
```
Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: **WARNING — Semantic coupling:** This reuses `_delayedExchangePluginIncompatibleVersion` (named for the delayed exchange plugin) to gate a conceptually independent feature (GlobalQos removal). Today both happen at 4.3, but if RabbitMQ ever re-enables GlobalQos or the threshold changes, this coupling becomes a bug. Consider introducing a dedicated `_globalQosRemovedVersion` constant (can reference the same `new Version(4, 3)` value — the point is semantic clarity).

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 2]
```diff
@@ -144,9 +144,22 @@ protected override async Task EnsureTopicSubscriptionAsync(CancellationToken can
...
+                {
+                    _logger.LogWarning("GlobalQos is not supported on RabbitMQ {ServerVersion}. Falling back to per-channel prefetch (global: false). Remove the GlobalQos option to suppress this warning", _serverVersion);
+                    useGlobalQos = false;
+                }
+                else if (useGlobalQos)
```
Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: **WARNING — False positive deprecation on unknown servers:** When `_serverVersion` is null (server didn't expose parseable version bytes), this `else if` branch fires unconditionally, telling the user GlobalQos "is deprecated in RabbitMQ 4.3+" even though we have no evidence the server is actually 4.3+. The user gets a deprecation warning for a feature that may work perfectly fine on their server.

Consider:
- `else if (useGlobalQos && _serverVersion is not null)` to only warn when we positively confirmed a pre-4.3 server, or
- Adjust the message to say something like "GlobalQos may be deprecated depending on your server version" when version is unknown.

**Review by niemyjski — COMMENTED** [review]
Self-review inline comments — see main review comment for full details.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Fixed. Introduced a dedicated \_globalQosRemovedVersion\ constant. Both point to \
ew Version(4, 3)\ today but are now semantically independent.

Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Fixed — same as above. Added \_globalQosRemovedVersion\ constant to decouple the two version checks.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 3]
```diff
@@ -591,7 +604,7 @@ protected override async Task PublishImplAsync(string messageType, object messag
         // if the RabbitMQ plugin is not available, then use the base class delay mechanism
         if (_delayedExchangePluginEnabled is false && options.DeliveryDelay.HasValue && options.DeliveryDelay.Value > TimeSpan.Zero)
         {
-            _logger.LogTrace("Schedule delayed message: {MessageType} ({Delay}ms)", messageType, options.DeliveryDelay.Value.TotalMilliseconds);
+            _logger.LogDebug("Delayed message will be scheduled in-memory (broker-side delayed exchange unavailable): {MessageType} ({Delay}ms)", messageType, options.DeliveryDelay.Value.TotalMilliseconds);
```
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Fixed. Reverted to \LogTrace\. The one-time warning at plugin detection time is sufficient for operators — the per-message log would be too noisy on 4.3+ where every delayed message takes this path.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 4]
```diff
@@ -144,9 +144,22 @@ protected override async Task EnsureTopicSubscriptionAsync(CancellationToken can
...
             {
-                await _subscriberChannel.BasicQosAsync(_options.PrefetchSize, _options.PrefetchCount, _options.GlobalQos, cancellationToken).AnyContext();
+#pragma warning disable CS0618 // GlobalQos is obsolete but we still need to read it for backward compatibility
+                bool useGlobalQos = _options.GlobalQos;
+#pragma warning restore CS0618
+                if (useGlobalQos && _serverVersion is not null && _serverVersion >= _delayedExchangePluginIncompatibleVersion)
+                {
+                    _logger.LogWarning("GlobalQos is not supported on RabbitMQ {ServerVersion}. Falling back to per-channel prefetch (global: false). Remove the GlobalQos option to suppress this warning", _serverVersion);
+                    useGlobalQos = false;
+                }
```
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Intentionally not adding a defensive try/catch. \_serverVersion\ is null in practice only on exotic non-RabbitMQ brokers (the AMQP handshake always includes server properties for RabbitMQ). If someone is on such a broker with \GlobalQos = true\, the channel error makes the problem immediately visible. The \[Obsolete]\ compile-time warning already steers users away from this option.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 5]
```diff
@@ -144,9 +144,22 @@ protected override async Task EnsureTopicSubscriptionAsync(CancellationToken can
...
             if (_options.PrefetchCount > 0 || _options.PrefetchSize > 0)
             {
-                await _subscriberChannel.BasicQosAsync(_options.PrefetchSize, _options.PrefetchCount, _options.GlobalQos, cancellationToken).AnyContext();
+#pragma warning disable CS0618 // GlobalQos is obsolete but we still need to read it for backward compatibility
+                bool useGlobalQos = _options.GlobalQos;
+#pragma warning restore CS0618
+                if (useGlobalQos && _serverVersion is not null && _serverVersion >= _delayedExchangePluginIncompatibleVersion)
+                {
+                    _logger.LogWarning("GlobalQos is not supported on RabbitMQ {ServerVersion}. Falling back to per-channel prefetch (global: false). Remove the GlobalQos option to suppress this warning", _serverVersion);
+                    useGlobalQos = false;
```
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs:154]: Good catch — updated the PR description to remove \silently\. The warning is intentional: operators should know the downgrade happened.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs (hunk 2 above)
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs]: Fixed. Gated the warning behind \_serverVersion is not null\ so we only warn when we positively know the server is pre-4.3. When version is unknown, the user gets no warning and the original \GlobalQos\ value is passed through unchanged.

**Review by niemyjski — COMMENTED** [review]
## Self-Review: PR #78 — Deprecate GlobalQos and delayed exchange plugin for RabbitMQ 4.3+

Self-review completed. Evaluated: security, correctness, performance, style. No blockers. The version-gated deprecation logic is sound and well-structured.

<details>
<summary>Full Review Details</summary>

### Security: PASS
No credential handling changes, no new dependencies, no build script modifications, no obfuscated code. Clean.

### Build: PASS | Tests: 53 passed (unit), 152 skipped (no local RabbitMQ) | CI: all 4 checks pass
- `dotnet build`: 0 warnings, 0 errors (net8.0 + net10.0)
- CI: build x2 pass, CodeQL pass, CLA pass

### Dependencies: N/A
No dependency changes.

### Requirements: ALIGNED
- ✅ `[Obsolete]` on `GlobalQos` property and builder method
- ✅ Runtime guard: `global: true` downgraded to `false` on 4.3+ with warning
- ✅ Pre-4.3 advisory deprecation warning (only when version is known)
- ✅ Delayed exchange log messages upgraded to Warning with clear fallback messaging
- ✅ Backward compatibility preserved — existing code compiles with warning only

### Code Review — Findings (severity-ranked)

**Medium-1: Pre-4.3 warning fires on every reconnection (Line 158)**

When connected to a known pre-4.3 server and `GlobalQos = true`, the code emits:
> "GlobalQos is deprecated in RabbitMQ 4.3+ and will be removed in a future version. Use per-channel prefetch instead"

This fires on every subscription setup (including reconnections). For operators on RabbitMQ 3.x who intentionally use `GlobalQos = true`, this is noise they cannot act on without changing their prefetch semantics. The `[Obsolete]` compile-time warning already covers the "you should migrate" message. Consider downgrading this runtime log to `LogInformation` or gating it behind a one-shot flag so it only fires once per instance lifetime.

Severity: **Medium** — Not a bug, but creates log noise for operators who can't migrate yet.

**Low-1: `_serverVersion` is not `volatile` (Line 34)**

Written in `DetectServerVersion` (called inside `_lock`) and read in `EnsureTopicSubscriptionAsync` (also inside `_lock`), so no data race in current code. However, if future changes read `_serverVersion` outside the lock, the lack of `volatile` could produce stale reads. Acceptable today given lock discipline.

Severity: **Low** — No current bug; defensive consideration only.

**Low-2: No unit test for the version-gated branching logic**

`ParseServerVersion` has thorough tests, but the three-way branch (4.3+ downgrade / pre-4.3 warning / null version passthrough) has no unit coverage. Acceptable since this is integration-level code requiring `IChannel` mocking, but would strengthen confidence.

Severity: **Low** — Simple branching logic, no complex state.

**Low-3: Repeated warning on delayed exchange detection (Lines 720-722)**

If publisher and subscriber race on first connection to a 4.3+ server, the delayed exchange warning could log twice before `_delayedExchangePluginEnabled` is set. No behavioral impact.

Severity: **Low** — Cosmetic log duplication only.

### PR-Level Assessment
- **Breaking changes:** None. `[Obsolete]` is additive (compile warning only).
- **Commit history:** Clean 2-commit progression: feature + fix.
- **Documentation:** PR body is thorough with test plan.
- **Existing comments:** All Copilot + self-review findings addressed in fix commit `c1a572f`.

### Existing Comments — Status
- Semantic coupling: ✅ Resolved — separate `_globalQosRemovedVersion` constant
- False deprecation on null version: ✅ Resolved — `_serverVersion is not null` guard
- LogTrace noise: ✅ Correctly addressed — kept at LogTrace with improved message
- "silently" in PR description: ✅ PR body updated
- Defensive try/catch: ✅ Intentionally declined with valid rationale

</details>

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 6]
```diff
@@ -144,9 +145,22 @@ protected override async Task EnsureTopicSubscriptionAsync(CancellationToken can
...
+                bool useGlobalQos = _options.GlobalQos;
+#pragma warning restore CS0618
+                if (useGlobalQos && _serverVersion is not null && _serverVersion >= _globalQosRemovedVersion)
+                {
+                    _logger.LogWarning("GlobalQos is not supported on RabbitMQ {ServerVersion}. Falling back to per-channel prefetch (global: false). Remove the GlobalQos option to suppress this warning", _serverVersion);
+                    useGlobalQos = false;
+                }
+                else if (useGlobalQos && _serverVersion is not null)
+                {
+                    _logger.LogWarning("GlobalQos is deprecated in RabbitMQ 4.3+ and will be removed in a future version. Use per-channel prefetch instead");
```
Comment (niemyjski) [inline src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs:158]: **Medium — Log noise on pre-4.3 servers:** This warning fires on every subscription setup (including reconnections). Operators on RabbitMQ 3.x who intentionally use \GlobalQos = true\ get repeated warnings they can't act on without changing prefetch semantics. The compile-time \[Obsolete]\ warning already tells them to migrate. Consider: (a) downgrading to \LogInformation\, or (b) using a one-shot flag (\private bool _globalQosWarningEmitted\) so it fires only once per instance lifetime. Not a blocker — just a quality-of-life consideration for operators mid-migration.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs [hunk 7]
```diff
@@ -144,9 +145,22 @@ protected override async Task EnsureTopicSubscriptionAsync(CancellationToken can
...
+                }
+
+                await _subscriberChannel.BasicQosAsync(_options.PrefetchSize, _options.PrefetchCount, useGlobalQos, cancellationToken).AnyContext();
                 _logger.LogDebug("Set channel QoS - PrefetchCount: {PrefetchCount}, PrefetchSize: {PrefetchSize}, Global: {GlobalQos} for acknowledgment strategy {AcknowledgementStrategy}",
-                    _options.PrefetchCount, _options.PrefetchSize, _options.GlobalQos, _options.AcknowledgementStrategy);
+                    _options.PrefetchCount, _options.PrefetchSize, useGlobalQos, _options.AcknowledgementStrategy);
             }
```
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs:164]: Acknowledged — the logic is simple three-way branching (version >= 4.3 → downgrade, version < 4.3 → warn, null → passthrough) that doesn't warrant mocking \IChannel\ for this PR. The integration tests against a real RabbitMQ 4.3 server (CI docker-compose) will exercise the 4.3+ path once we upgrade the test container. Adding a unit test for version comparison arithmetic provides minimal confidence beyond what the code already expresses. Won't add for this PR.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs:158]: The wording is intentionally simplified for operator clarity. RabbitMQ 4.3+ moved `global_qos` to `denied_by_default`, which means it's effectively "not supported" unless explicitly re-enabled in server config — something most operators won't do. Saying "denied by default" in a log message would require operators to look up what that means. The current message is actionable: remove the option, problem solved.

The pre-4.3 branch message is accurate as an advisory: it tells operators their current config works today but won't survive a 4.3 upgrade. This is the point — give them migration notice while things still work.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBusOptions.cs [hunk 8]
```diff
@@ -71,6 +71,7 @@ public class RabbitMQMessageBusOptions : SharedMessageBusOptions
     /// When true, the QoS settings apply to all consumers on the connection.
     /// When false (default), the QoS settings apply only to consumers on this channel.
     /// </summary>
+    [Obsolete("Global QoS is deprecated in RabbitMQ 4.3+ and will be removed in a future version. Use per-channel prefetch (GlobalQos = false) instead.")]
```
[the remark this answers is not in the document]
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBusOptions.cs:74]: Keeping "deprecated" rather than "denied by default" — the `[Obsolete]` attribute is a compile-time signal aimed at library consumers, not server admins. "Deprecated" accurately describes the lifecycle status from the consumer's perspective: this option is going away in a future version of the library regardless of whether the server permits it. The implementation detail of how RabbitMQ 4.3 enforces it (denied_by_default vs. fully removed) isn't relevant to the consumer deciding whether to use this API.

File: src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs (hunk 6 above)
Comment (niemyjski) [reply src/Foundatio.RabbitMQ/Messaging/RabbitMQMessageBus.cs:158]: Agree it's a quality-of-life consideration but leaving as-is for this PR. Rationale:

1. The warning is **actionable**: remove `GlobalQos = true` from config and it disappears
2. Operators mid-migration benefit from seeing it on reconnect — confirms the option is still set after deploys/restarts
3. The `[Obsolete]` compile-time warning handles the "developer awareness" path
4. Adding a one-shot flag adds state complexity for a warning that only fires when users opt into a deprecated feature

If this generates noise in practice, we can downgrade to `LogInformation` in a follow-up. Not blocking this PR on it.
